### PR TITLE
Refactoring interaction

### LINKED ISSUE #14: Refactoring Interactions
## Description

Plan to refactor the interaction with plots:
- Merge Interaction classes with AxesModier classes
- Separate interaction classes for each specific plot
- Explore the option of using parent classes
- Explore the option of attaching the Interaction/Modifier to the axis 
- Should avoid length `if ... else` checks
- More modular code
- Log/Lin x-axes via `set_xscale`

Related to: #21

### LINKED ISSUE #21: Plot interaction hotkeys config file
## Description

Move the configuration of hotkeys for interactions with plots into a config file rather than hard-coding them.
Options could be:
- *.ini style configs
- *.json style
Internally we could use a dict.

Related to: #14

### PR DESCRIPTION
**Pull request by f-brinkmann**
### Which issue(s) are closed by this pull request?

closes #14, closes #21

### Note
This is a fully working draft but the new Interaction class in only used if calling `plot.line.time`. Current keyboard shortcuts are listed in `plot.utils.shortcuts`.

### Changes proposed in this pull request:

**Completely refactored interaction module**
- added `EventEmu` class to emulate key events for debugging and class internal use
- added `PlotParameter` class for remembering user parameters and setting plot depending axis parameters
- `Cycle` class only cycles the index, because the data that is cycled depends on the plot type. The data for the current index is now accessed inside `Interaction`

**Structural changes in `Interaction` class**
- operation should be more clear - a central `select_action()` method routes the keyboard input to all other action functions
- Implemented AxisModifiers as part of Interaction.
- There is only one Axis modifier function that can handle all axis types: `apply_move_and_zoom`

**Functional changes in the `Interaction` class**
- using keyboard shortcuts from `plot.utils.shortcuts`
- Read keyboard shortcuts from json file
- More than one key can be assigned to an action
- remember user settings (e.g., the time signal being plotted in dB)
- improved move and zoom operations for different axes types (lin, dB, frequency)
- zooming and moving colorbar ranges
- zooming and moving planar plots such as the `spectrogram`
- Showing channel numbers when cycling though plots
- remember the currently active channel when toggling between plots
- always show channel number for spectrogram, if the signal contains more than one channel
- full support of 'light' and 'dark theme'

### Not done as discussed
- Mouse over activation of interaction in case of subplot-layouts

### Next step in a follow up pull-request
- Add basic testing. There was no testing for the old Interaction class. Testing might be done by using the `EventEmu` class to generate baseline and test plots. Testing will be limited to a reasonable amount.

### CODE REVIEW

**Review by sikersten — APPROVED** [review]
Thanks very much for your effort on implementing these interactions! Only detected little malfunctions
* x-axis zoom on a logarithmic frequency axis leads to additional ticks of scientific style causing the overall tick style to be mixed and the label to move
* Not actually part of this pull request, but resizing the window changes the axes-to-window distances. Could this be solved with an matplotlib ResizeEvent in `select_action()`?

**Comment by f-brinkmann** [comment]
>     * x-axis zoom on a logarithmic frequency axis leads to additional ticks of scientific style causing the overall tick style to be mixed and the label to move
> 
>     * Not actually part of this pull request, but resizing the window changes the axes-to-window distances. Could this be solved with an matplotlib ResizeEvent in `select_action()`?

I fixed the moving label. I think the ticks are an issue of the log ticker, not interaction (I opened a new issue). I am not sure if I understand the second point - let's discuss during the next weekly :)

**Review by mberz — COMMENTED** [review]
I have tested the interactions for most plots now, so far I had issues with the toggle functions for axes types (log vs. lin) which seemed to be not working, as well as the functionality that is specific to plots involving a colormap. Since this is a draft, I've not ventured further into this.

Btw.: What do you think of including cycling the channels on spectrogram plots? I'd love to see that feature.

One more important thing that comes to mind: The logic has become rather complex and is not very easily digestible for someone who is not familiar with this. Could you include some sort of table/diagram (or similar) which outlines which interaction is usable on which plot? This could be very helpful for future develpers/maintainers.

**Comment by f-brinkmann** [comment]
> 
> 
> I have tested the interactions for most plots now, so far I had issues with the toggle functions for axes types (log vs. lin) which seemed to be not working, as well as the functionality that is specific to plots involving a colormap. Since this is a draft, I've not ventured further into this.
> 
> Btw.: What do you think of including cycling the channels on spectrogram plots? I'd love to see that feature.
> 
> One more important thing that comes to mind: The logic has become rather complex and is not very easily digestible for someone who is not familiar with this. Could you include some sort of table/diagram (or similar) which outlines which interaction is usable on which plot? This could be very helpful for future develpers/maintainers.

- Can you post the error you get when trying to toogle the axes?
- Cycling channels should already be supported. Please let me know if it does not work for you.
- `plots.utils.shortcuts` now also lists, what shortcuts are available for which plot function

File: pyfar/plot/_line.py [hunk 1]
```diff
@@ -254,10 +254,13 @@ def _phase(signal, deg=False, unwrap=False, xscale='log', ax=None, **kwargs):
...
         ylabel_string += 'in radians'
-        ax.yaxis.set_major_locator(MultipleFractionLocator(np.pi, 2))
-        ax.yaxis.set_minor_locator(MultipleFractionLocator(np.pi, 6))
-        ax.yaxis.set_major_formatter(MultipleFractionFormatter(
-            nominator=1, denominator=2, base=np.pi, base_str=r'\pi'))
+        # nice tick formatting is not done for unwrap=True. In this case
+        # it can create 1000 or more ticks.
+        if not unwrap or unwrap == "360":
+            ax.yaxis.set_major_locator(MultipleFractionLocator(np.pi, 2))
+            ax.yaxis.set_minor_locator(MultipleFractionLocator(np.pi, 2))
```
[the remark this answers is not in the document]
Comment (f-brinkmann) [reply pyfar/plot/_line.py:261]: thanks for noting. Corrected the minor ticks to be as before.

File: pyfar/plot/line.py [hunk 2]
```diff
@@ -54,9 +55,13 @@ def time(signal, dB=False, log_prefix=20, log_reference=1, unit=None, ax=None,
...
     plt.tight_layout()
-    Interaction('LineXLin', ax, signal, style, **kwargs)
+    plot_parameter = ia.PlotParameter(
+        'line.time', dB_time=dB, log_prefix=log_prefix,
+        log_reference=log_reference)
+    interaction = ia.Interaction(
+        signal, ax, style, plot_parameter, **kwargs)
 
-    return ax
+    return ax, interaction
```
[the remark this answers is not in the document]
Comment (f-brinkmann) [reply pyfar/plot/line.py]: That was only for debugging purposes. Now the interaction is attached to the axes for all plots.

File: pyfar/plot/utils.py [hunk 3]
```diff
@@ -94,42 +94,99 @@ def shortcuts(show=True):
     short_cuts = {
```
[the remark this answers is not in the document]
Comment (f-brinkmann) [reply pyfar/plot/utils.py]: This was on the ToDo List and is now possible :)

File: pyfar/plot/interaction.py [hunk 4]
```diff
@@ -0,0 +1,749 @@
...
+        changed = False
+        if self.y_type is not None:
+            if len(self._y_type) > 1:
+                self._y_id = (self._y_id + 1) % len(self._y_type)
+                setattr(self, self._y_param, self._y_values[self._y_id])
+                changed = True
+
+        return changed
+
+    def toggle_cm(self):
```
[the remark this answers is not in the document]
Comment (f-brinkmann) [reply pyfar/plot/interaction.py]: renamed it to `toggle_colormap` and gave an example in the docstring to make this more clear. Also added examples to the remaining `toggle_*` functions.

File: pyfar/plot/interaction.py [hunk 5]
```diff
@@ -0,0 +1,749 @@
...
+
+        # get new limits
+        lims_new = lims + shift
+
+        # apply limits
+        setter_function(lims_new[0], lims_new[1])
+
+        self.draw_canvas()
+
+    def toggle_all_lines(self):
```
[the remark this answers is not in the document]
Comment (f-brinkmann) [reply pyfar/plot/interaction.py]: I checked the release note and I am still not sure what this is about. Let's wait what will happen...

**Comment by mberz** [comment]
Comments above seem to be caused by faulty conda env. Merging after everything has been resolved.

**Pull request merged**